hal/ia32/cpu.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * CPU related routines
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+#include <syspage.h>
+
+
+/* FIXME: temporary usage until new syspage implementation */
+extern int hal_memoryGetEntry(unsigned int *offs, mem_entry_t *entry);
+
+
+__attribute__((noreturn)) void hal_cpuJump(addr_t addr)
+{
+	/* FIXME: temporary HAL syspage usage, should be removed after new syspage implementation is added */
+	void *syspage = (void *)syspage_getAddress();
+	unsigned short *mmsz = syspage + 300;
+	mem_entry_t *mm = (mem_entry_t *)(mmsz + 1);
+	unsigned short *progssz = (unsigned short *)(mm + 64);
+	unsigned int offs = 0;
+
+	/* Copy HAL syspage data */
+	syspage_setHalData(&_plo_syspage);
+	syspage_save();
+
+	/* Add memory map entries */
+	*mmsz = 0;
+	do {
+		if (hal_memoryGetEntry(&offs, mm))
+			break;
+		(*mmsz)++;
+		mm++;
+	} while (offs);
+	hal_memset(mm, 0, (64 - *mmsz) * sizeof(mem_entry_t));
+
+	/* Add program entries */
+	*progssz = 0;
+
+	/* Generate fake syspage data - required for current kernel syspage compatibility */
+	*(unsigned int *)(syspage + 0x20) = *(unsigned int *)(syspage + sizeof(syspage_hal_t)); /* Kernel */
+	*(unsigned int *)(syspage + 0x24) = 0xc0000;                                            /* Kernel size */
+	*(unsigned int *)(syspage + 0x28) = 0x0;                                                /* Console */
+	*(char *)(syspage + 0x2c) = '\0';                                                       /* Args */
+
+	hal_done();
+
+	__asm__ volatile(
+		"cli; "
+		"movl 24(%0), %%esp; "
+		"pushl %0; "
+		"jmpl *%1; "
+	:: "g" (syspage), "r" (addr));
+
+	__builtin_unreachable();
+}
+
+
+void hal_cpuHalt(void)
+{
+	__asm__ volatile("hlt":);
```

Comment:
__asm__ volatile("hlt"); should be ok

Reply:
GCC asm documentation says that extended asm (with ':') is preffered to basic asm (without ':') ([link, remarks section](https://gcc.gnu.org/onlinedocs/gcc/Basic-Asm.html#Basic-Asm)):
> Using extended asm (see Extended Asm) typically produces smaller, safer, and more efficient code, and in most cases it is a better solution than basic asm. However, there are two situations where only basic asm can be used:
> 1. Extended asm statements have to be inside a C function, so to write inline assembly language at file scope (“top-level”), outside of C functions, you must use basic asm. You can use this technique to emit assembler directives, define assembly language macros that can be invoked elsewhere in the file, or write entire functions in assembly language. Basic asm statements outside of functions may not use any qualifiers.
> 2. Functions declared with the naked attribute also require basic asm (see Function Attributes). 

Also using `__asm__` instead of `asm` is safer when using specific compilation options:
> The `asm` keyword is a GNU extension. When writing code that can be compiled with -ansi and the various -std options, use `__asm__` instead of `asm`

All of the inline asm in this PR is written using extended asm and I'd like to keep it that way.